packages/drivers/odsp-driver/src/getFileLink.ts

```diff
@@ -95,6 +105,48 @@ export async function getFileLink(
 	return fileLink;
 }
 
+/**
+ * Handles location redirection while fulfilling the getFilelink call. We don't want browser to handle
+ * the redirection as the browser will retry with same auth token which will not work as we need app
+ * to regenerate tokens for the new site domain.
+ * @param getToken - token fetcher to fetch the token.
+ * @param odspUrlParts - parts of odsp resolved url.
+ * @param logger - logger to send events.
+ * @returns Response from the API call.
+ * @alpha
+ */
+async function getFileLinkWithLocationRedirectionHandling(
+	getToken: TokenFetcher<OdspResourceTokenFetchOptions>,
+	odspUrlParts: IOdspUrlParts,
+	logger: ITelemetryLoggerExt,
+): Promise<string> {
+	// We can have chains of location redirection one after the other, so have a for loop
+	// so that we can keep handling the same type of error.
+	for (;;) {
+		try {
+			return await getFileLinkCore(getToken, odspUrlParts, logger);
+		} catch (error: unknown) {
+			if (
+				isFluidError(error) &&
+				error.errorType === OdspErrorTypes.fileNotFoundOrAccessDeniedError
+			) {
+				const redirectLocation = (error as IOdspErrorAugmentations).redirectLocation;
+				if (redirectLocation !== undefined) {
+					logger.sendTelemetryEvent({
+						eventName: "LocationRedirectionErrorForGetOdspFileLink",
+					});
+					// Generate the new SiteUrl from the redirection location.
+					const newSiteDomain = new URL(redirectLocation).origin;
+					const newSiteUrl = `${newSiteDomain}${new URL(odspUrlParts.siteUrl).pathname}`;
+					odspUrlParts.siteUrl = newSiteUrl;
+					continue;
```

Comment:
Any easy test we can make that it does not result in pointing back to previous domain?
I wonder if someone could create a circular link here :)
I.e. can malicious user create a circular link and send it to hundreds of people, causing DDOS attack on SPO?

Reply:
Maybe @AndreiZe comment would work in this case where he suggested that we can max limit on it for redirections. 
Not sure if the user can change the domain of file by themselves or it is done by service.

Reply:
Added max retries.